### DISCUSSION TITLE
Probably incorrect variance of type variables in PEP-695 syntax with frozen dataclasses.

### DISCUSSION THREAD

**Discussion by tkukushkin** [microsoft/pyright#5568] [Ideas]
I try to use pyright with dev version of python 3.12.

For example I have such code:

```python
class Foo[T]:
    def __init__(self, x: T) -> None:
        self._x = x

    @property
    def x(self) -> T:
        return self._x

x: Foo[int] = Foo(1)
y: Foo[float] = x
```

Last assignment is ok for pyright, it properly derives variance of `T` as covariant.

But if I want to use frozen dataclass here instead, pyright derives `T` as invariant and shows error on assignment:

```python
from dataclasses import dataclass

@dataclass(frozen=True)
class Foo[T]:
    x: T

x: Foo[int] = Foo(1)
y: Foo[float] = x  # Expression of type "Foo[int]" cannot be assigned to declared type "Foo[float]"
```

I am not sure if it's bug or not, but with the current behavior it's not very convenient.

**Comment by erictraut** [comment]
I agree that the variance should be inferred as covariant in this case because the attribute is read-only. I've added a fix, which will be in the next release.

**Reply by tkukushkin** [reply]
I thought about it more and now I don't understand, how inference of variance works in more complicated situations.

For example pyright doesn't see any problems with this code:

```python
class Foo[T]:
    def __init__(self, x: T) -> None:
        self.x = x


def foo(x: Foo[float]) -> None:
    x.x = 2.0

x: Foo[int] = Foo(1)
foo(x)
```

Pyright infers `Foo` as covariant, but actually `x` attribute is not read-only. And I don't see any ways to tell type checker that my class is **not** covariant using new 3.12 syntax.

**Reply by erictraut** [reply]
You've uncovered another bug. The variance inference code in pyright was ignoring instance-only variables because the algorithm was originally written for protocols, and protocols cannot have instance-only variables. I've fixed the problem, and it will be addressed in the next release. If you see other issues like this, please feel free to open bug reports.

**Reply by tkukushkin** [reply]
So, if it is a bug, the class should be invariant? 

And then this class should be invariant too?

```python
class Foo[T]:
    def __init__(self, x: T) -> None:
        self._x = x

    @property
    def x(self) -> T:
        return self._x
```

But if we consider `_x` as protected field, which shouldn't be accessed outside of class and it's inheritors, then property `x` is read-only, as with frozen dataclass, and class should be covariant. How does pyright handle it?

Maybe it is not a problem of pyright, I am just trying to understand the rules, how it should work.

**Reply by erictraut** [reply]
A private or protected field is an interesting case that I hadn't considered. Those should probably be exempt from the invariance inference algorithm. Do you agree?

**Reply by tkukushkin** [reply]
I probably agree, but my opinion doesn't mean much here :-)
